WordPress/gutenberg
# Add custom social link icons and use variations

## What?
Closes #19041.

This PR changes the social-link block to use the variations icon/label in the editor and adds a PHP filter to adding custom social services. 
It's similar to [ndiegos social sharing link block](https://github.com/ndiego/social-sharing-block/pull/23). 



## Why?


- add custom icons - maybe for social networks that are regional and not globally available or just a niche
- core doesn't have to provide all possible icons there are 
- change icons to fit the style of the website
- have a good editing experience = show the same icon in the editor as on the frontend

Related discussions/PR: 
#39419 
#30749
#39868
#55142

## How?

- use active variation data in editor (variations already have an icon + label defined)
- add filter in php to allow adding services there ( + cache them since they can be called multiple times)

## TODO 
- [ ] Currently the filter name gets prefixed with `gutenberg_`  in the build process and I don't know why.
- [ ] make changes in edit.native.js

## Testing Instructions
### Existing social blocks
1. Add a social links block
2. Add a facebook link block and see the facebook logo being used in the editor + on the frontend

### Custom social icons
In a custom plugin/theme
1. Register a variation with an icon in the editor. Dashicon-Slugs as well as SVG Icon Components (eg from @wordpress/icons) can be used:
```js
registerBlockVariation("core/social-link", {
  name: "custom-social-service-2",
  title: "Custom Social Service 2",
  description: "A link to a custom social service with a custom SVG icon",
  icon: createElement(
    SVG,
    {
      viewBox: "0 0 24 24",
      xmlns: "http://www.w3.org/2000/svg",
      "aria-hidden": "true",
      focusable: "false",
    },
    createElement(Path, {
      d: "M18 4H6c-1.1 0-2 .9-2 2v12.9c0 .6.5 1.1 1.1 1.1.3 0 .5-.1.8-.3L8.5 17H18c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2zm.5 11c0 .3-.2.5-.5.5H7.9l-2.4 2.4V6c0-.3.2-.5.5-.5h12c.3 0 .5.2.5.5v9z",
    })
  ),
  attributes: {
    service: "custom-social-service-2",
  },
  isActive: ["service"],
});
```
- test existing social blocks
- test cusotm social blocks with plugin 




2. Open Editor and add the custom social service to a social links block. The Dashicon for post-format (a speech-bubble) should be shown
3. Add PHP for the frontend: 
```php
add_action('enqueue_block_editor_assets', function () {
    wp_enqueue_script(
        'social-link-variation',
        plugins_url('index.js', __FILE__),
        ['wp-element', 'wp-primitives', 'wp-blocks', 'wp-components'],
        filemtime(plugin_dir_path(__FILE__, 'index.js')),
    );
});

function my_add_custom_social_link_data($services_data)
{
    $services_data['custom-social-service'] = [
        'name' => 'Custom Social Service',
        'icon' => '<span class="dashicons dashicons-format-status"></span>'
    ];
    $services_data['custom-social-service-2'] = [
        'name' => 'Custom Social Service 2',
        'icon' => '<svg viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false"><path d="M18 4H6c-1.1 0-2 .9-2 2v12.9c0 .6.5 1.1 1.1 1.1.3 0 .5-.1.8-.3L8.5 17H18c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2zm.5 11c0 .3-.2.5-.5.5H7.9l-2.4 2.4V6c0-.3.2-.5.5-.5h12c.3 0 .5.2.5.5v9z"></path></svg>'
    ];
    return $services_data;
}

add_filter('gutenberg_block_core_social_link_services', 'my_add_custom_social_link_data');
add_filter('block_core_social_link_services', 'my_add_custom_social_link_data');
```
4. View post/page on the frontend, the icon should be displayed on the frontend as well. 

### Testing Instructions for Keyboard


## Screenshots or screencast
